# Task 1

# Predict Restaurant Ratings

In [17]:
import pandas as pd

# Load the dataset
file_path = ("C:\\Users\\trupti kadam\\Downloads\\Dataset .csv")
df = pd.read_csv(file_path, encoding='UTF-8-SIG')

# Display the first few rows of the dataframe
df.head()

Restaurant ID         Restaurant Name  Country Code              City  \
0        6317637        Le Petit Souffle           162       Makati City   
1        6304287        Izakaya Kikufuji           162       Makati City   
2        6300002  Heat - Edsa Shangri-La           162  Mandaluyong City   
3        6318506                    Ooma           162  Mandaluyong City   
4        6314302             Sambo Kojin           162  Mandaluyong City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2  Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3  Third Floor, Mega Fashion Hall, SM Megamall, O...   
4  Third Floor, Mega Atrium, SM Megamall, Ortigas...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   
2  Edsa Shangri-La, Ortigas, Mandaluyong City   
3      SM Megamall, Ortigas, Mandaluyong City   
4      SM Megamall, Ortigas, Mandaluyong City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   
2  Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831  14.581404   
3  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475  14.585318   
4  SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508  14.584450   

                           Cuisines  ...          Currency Has Table booking  \
0        French, Japanese, Desserts  ...  Botswana Pula(P)               Yes   
1                          Japanese  ...  Botswana Pula(P)               Yes   
2  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)               Yes   
3                   Japanese, Sushi  ...  Botswana Pula(P)                No   
4                  Japanese, Korean  ...  Botswana Pula(P)               Yes   

  Has Online delivery Is delivering now Switch to order menu Price range  \
0                  No                No                   No           3   
1                  No                No                   No           3   
2                  No                No                   No           4   
3                  No                No                   No           4   
4                  No                No                   No           4   

   Aggregate rating  Rating color Rating text Votes  
0               4.8    Dark Green   Excellent   314  
1               4.5    Dark Green   Excellent   591  
2               4.4         Green   Very Good   270  
3               4.9    Dark Green   Excellent   365  
4               4.8    Dark Green   Excellent   229  

[5 rows x 21 columns]

In [18]:
# Check for missing values
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [19]:
# Fill missing values in 'Cuisines' with 'Unknown'
df['Cuisines'].fillna('Unknown', inplace=True)

# Confirm if missing values are handled
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                0
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Encoding categorical variables
encoder = LabelEncoder()
categorical_columns = ['City', 'Cuisines', 'Currency', 'Has Table booking', 'Has Online delivery', 'Is delivering now', 'Switch to order menu', 'Rating color', 'Rating text']
for column in categorical_columns:
    df[column] = encoder.fit_transform(df[column])

# Splitting the data into training and testing sets
X = df.drop(['Restaurant ID', 'Restaurant Name', 'Country Code', 'Address', 'Locality', 'Locality Verbose', 'Aggregate rating'], axis=1)
y = df['Aggregate rating']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print('Training set size:', X_train.shape[0])
print('Testing set size:', X_test.shape[0])

Training set size: 7640
Testing set size: 1911


In [21]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Split data into X and y
X = df.drop('Aggregate rating', axis=1)
y = df['Aggregate rating']

# Linear Regression
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)
y_pred = lin_reg.predict(X_test)

lin_mse = mean_squared_error(y_test, y_pred)
lin_r2 = r2_score(y_test, y_pred)

print(f'Linear Regression MSE: {lin_mse:.3f}')
print(f'Linear Regression R-squared: {lin_r2:.3f}')

# Decision Tree Regression 
dtree_reg = DecisionTreeRegressor()
dtree_reg.fit(X_train, y_train)
y_pred = dtree_reg.predict(X_test)

dtree_mse = mean_squared_error(y_test, y_pred) 
dtree_r2 = r2_score(y_test, y_pred)

print(f'Decision Tree Regression MSE: {dtree_mse:.3f}')
print(f'Decision Tree Regression R-squared: {dtree_r2:.3f}')

Linear Regression MSE: 1.351
Linear Regression R-squared: 0.406
Decision Tree Regression MSE: 0.054
Decision Tree Regression R-squared: 0.976


The regression models have been trained and evaluated on the testing data with the following results:

Linear Regression:

Mean Squared Error (MSE): 1.351
R-squared: 0.406
Decision Tree Regression:

Mean Squared Error (MSE): 0.054
R-squared: 0.976
The Decision Tree Regression model outperformed the Linear Regression model significantly, with a much lower MSE and a higher R-squared value. This indicates that the Decision Tree model better captures the variance in the data and provides more accurate predictions of restaurant ratings.